tiledb/api/src/stats.rs

```diff
@@ -0,0 +1,70 @@
+use crate::error::Error;
+use crate::Result as TileDBResult;
+
+pub fn stats_enable() -> TileDBResult<()> {
```

Comment:
I'll offer the disclaimer that I know absolutely nothing about what this is doing inside core.

I think this is a little too C-like.  Maybe it's not a safety or correctness issue but it does seem like a performance issue if you turned stats on and then bad control flow forgot to turn them off.  I'd probably want this function to give me a struct back, whose `impl Drop` turns the stats collection off.  And maybe this struct has a method that gives you the stats back, instead of a raw function.

That doesn't solve all the problems, of course - it's still not quite right if you put multiples of them in your scope - but it will be more user-friendly than a 1:1 mapping between C and Rust functions

Reply:
A slight disagreement, I'd add the ability to do what you're asking, but leave these APIs. Sometimes you really do just want to enable stats so that you can report periodically. So basically, keep this API shape as is, but add a `with_stats_enabled()` API that returns some struct/enum/whatever that has a drop trait that disables things.

Although, it suddenly occurs to me to ask what happens if a drop trait fails? In this case its probably super unlikely but it might be something to read up on. I'm guessing panic land but who knows.

Reply:
If the overhead of stats colleciton is noticeable, and this is eventually running in a multi-user environment, then there is definitely a risk that user A will turn on stats collection and forget to turn it off, and user B will yell at TileDB support that their queries are suddenly slower but nothing changed.

If user A wants to have stats enabled for all their workloads then core must support having stats collected at a better granularity than global. And for our API, maybe user A would make this thing a const.  That is an option (I think), to make whatever the structure is const-constructible.

Since Rust doesn't really have exception handling the only way for anything to fail is to return an error or to panic, and since Drop doesn't have a return type it can only panic.  This particular case doesn't really seem panic appropriate to me (I think the worst-case scenario is more or less what I outlined above, and maybe some disk usage quota issues if stats does any logging) but logging it as an error would be wise.

I guess that means we should start thinking about logging.  Oh boy.  Not for this PR though

Reply:
Yeah, I think maybe the existence of both APIs (returning an object that automatically disables stats when it goes out of scope (effectively RAII stats collection) and the normal API works well. I'll add that!